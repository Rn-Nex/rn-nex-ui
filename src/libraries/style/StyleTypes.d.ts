import {
  AnimatableNumericValue,
  DimensionValue,
  FlexStyle,
  ImageStyle,
  TextProps as RnTextProps,
  TextStyle,
  TouchableOpacity,
  TouchableOpacityProps,
  TouchableWithoutFeedback,
  TransformsStyle,
  View,
  ViewProps,
  ViewStyle,
} from 'react-native';
import { BoxProps } from '../../components/Box/interfaces';
import { TextProps } from '../../components/Typography/interfaces';

/**
 * Interface defining properties related to spacing styles, such as margins and paddings.
 * This extends the FlexStyle interface and picks specific properties relevant to spacing.
 */
export interface SpacingStyle
  extends Pick<
    FlexStyle,
    | 'gap'
    | 'margin'
    | 'marginBottom'
    | 'marginEnd'
    | 'marginHorizontal'
    | 'marginLeft'
    | 'marginRight'
    | 'marginStart'
    | 'marginTop'
    | 'marginVertical'
    | 'padding'
    | 'paddingBottom'
    | 'paddingEnd'
    | 'paddingHorizontal'
    | 'paddingLeft'
    | 'paddingRight'
    | 'paddingStart'
    | 'paddingTop'
    | 'paddingVertical'
  > {}

export interface StylePalette
  extends ViewStyle,
    TextStyle,
    TransformsStyle,
    ImageStyle,
    ElementSpacingMap,
    ElementBorderRadiusMap {}

export type KeyOfStylePalette = keyof StylePalette;
export type KeyOfTextStyleFontWeight = keyof TextStyle['fontWeight'];

/**
 * Define the object structure for each entry in the array Property name should be one of
 * the keys of StylePalette
 * Value type is determined based on the propertyName
 */
export interface StyleEntry<T extends KeyOfStylePalette> {
  propertyName: T;
  value: StylePalette[T];
}

/**
 * Define a type representing the props of common native elements such as View, Text, TouchableOpacity.
 */
export type NativeElementProps = ViewProps | RnTextProps | TouchableOpacityProps;

/**
 * Define a type representing the props of styled components.
 */
export type StyledElementProps = BoxProps | TextProps;

/**
 * Define a type representing common native elements used in the application,
 * including View, Text, TouchableOpacity, and TouchableWithoutFeedback. etc.
 */
export type NativeElements = View | Text | TouchableOpacity | TouchableWithoutFeedback;

/**
 * Define a type representing React component ref attributes for the native elements, ensuring
 * compatibility with React's ref system.
 */
export type ReactComponentRefAttributesType = React.RefAttributes<NativeElements>;
/**
 * Interface defining dimensions for width and height.
 * Each dimension can be a number, string, or a combination of both (e.g., '100%', 200).
 */
export interface DimensionInterface {
  width?: DimensionValue;
  height?: DimensionValue;
}
/**
 * Defines the types of padding that can be applied to an element.
 * @example
 * 'p': Padding on all sides
 * 'px': Horizontal padding (left and right)
 * 'py': Vertical padding (top and bottom)
 * 'ps': Padding on the start side (left in LTR, right in RTL)
 * 'pe': Padding on the end side (right in LTR, left in RTL)
 * 'pt': Padding on the top side.
 * 'pb': Padding on the bottom side
 */
export type ElementPaddingTypes = 'p' | 'px' | 'py' | 'ps' | 'pe' | 'pt' | 'pb';

/**
 * Defines the types of margin that can be applied to an element.
 * @example
 * 'm': Margin on all sides
 * 'mx': Horizontal margin (left and right)
 * 'my': Vertical margin (top and bottom)
 * 'ms': Margin on the start side (left in LTR, right in RTL)
 * 'me': Margin on the end side (right in LTR, left in RTL)
 * 'mt': Margin on the top side
 * 'mb': Margin on the bottom side
 */
export type ElementMarginTypes = 'm' | 'mx' | 'my' | 'ms' | 'me' | 'mt' | 'mb';

/**
 * Represents a mapping of padding types to their corresponding dimension values.
 * Each key in the ElementPaddingTypes represents a padding type, and the value is a DimensionValue.
 * DimensionValue can be a string representing a CSS size (e.g., '10px') or a number representing pixels.
 */
export type ElementPaddingMap = {
  [key in ElementPaddingTypes]?: DimensionValue;
};
/**
 * Represents a mapping of margin types to their corresponding dimension values.
 * Each key in the ElementMarginTypes represents a margin type, and the value is a DimensionValue.
 * DimensionValue can be a string representing a CSS size (e.g., '10px') or a number representing pixels.
 */
export type ElementMarginMap = {
  [key in ElementMarginTypes]?: DimensionValue;
};

/**
 * Represents a mapping of padding and margin types to their corresponding DimensionInterface.
 * Each key in the ElementSpacingMap represents either a padding type or a margin type,
 * and the value is a DimensionInterface defining the dimension properties (e.g., width, height).
 */
export type ElementSpacingMap = {
  [key in ElementPaddingTypes | ElementMarginTypes]?: DimensionValue;
};
export interface SpacingProps extends ElementSpacingMap, SpacingStyle {}

/**
 * interface for element border radius properties, which may be used in styling
 */
export interface ElementRadius
  extends Pick<
    ViewStyle,
    | 'borderRadius'
    | 'borderTopEndRadius'
    | 'borderTopLeftRadius'
    | 'borderEndEndRadius'
    | 'borderTopRightRadius'
    | 'borderEndStartRadius'
    | 'borderStartEndRadius'
    | 'borderTopStartRadius'
    | 'borderBottomEndRadius'
    | 'borderBottomLeftRadius'
    | 'borderStartStartRadius'
    | 'borderBottomRightRadius'
    | 'borderBottomStartRadius'
    | 'shadowRadius'
  > {}
/**
 * Type alias for element border radius properties.
 * Abbreviations:
 * - 'r': 'borderRadius'
 * - 'ret': 'borderTopEndRadius'
 * - 'rlt': 'borderTopLeftRadius'
 * - 'ree': 'borderEndEndRadius'
 * - 'rrt': 'borderTopRightRadius'
 * - 'rse': 'borderEndStartRadius'
 * - 'res': 'borderStartEndRadius'
 * - 'rst': 'borderTopStartRadius'
 * - 'reb': 'borderBottomEndRadius'
 * - 'rlb': 'borderBottomLeftRadius'
 * - 'rss': 'borderStartStartRadius'
 * - 'rrb': 'borderBottomRightRadius'
 * - 'rsb': 'borderBottomStartRadius'
 */
export type ElementBorderRadius =
  | 'r'
  | 'ret'
  | 'rlt'
  | 'ree'
  | 'rrt'
  | 'rse'
  | 'res'
  | 'rst'
  | 'reb'
  | 'rlb'
  | 'rss'
  | 'rrb'
  | 'rsb';

/**
 * Mapping of element border radius keys to their respective animatable numeric values.
 */
export type ElementBorderRadiusMap = {
  [key in ElementBorderRadius]?: AnimatableNumericValue;
};

export interface ElementBorderColorStyles
  extends Pick<
    ViewStyle,
    'borderBlockColor' | 'borderBlockEndColor' | 'borderBlockStartColor' | 'borderBottomColor' | 'borderColor'
  > {}

export interface ElementBorderStyles extends ElementRadius, ElementBorderColorStyles {}
