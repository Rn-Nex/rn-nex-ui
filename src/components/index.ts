export * from './Accordion';
export * from './ActivityIndicator';
export * from './Avatar';
export * from './Badge';
export * from './Box';
export * from './Button';
export * from './Card';
export * from './CheckBox';
export * from './Chip';
export * from './Dialog';
export * from './Divider';
export * from './DropDown';
export * from './Image';
export * from './List';
export * from './Pagination';
export * from './Portal';
export * from './Radio';
export * from './Snackbar';
export * from './Stepper';
export * from './Switch';
export * from './TextField';
export * from './Transitions';
export * from './Typography';
