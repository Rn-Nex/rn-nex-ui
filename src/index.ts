export * from './components';
export * from './libraries';
export * from './utils';
