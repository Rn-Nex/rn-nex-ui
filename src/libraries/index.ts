export * from './themes/v2';
export * from './style';
