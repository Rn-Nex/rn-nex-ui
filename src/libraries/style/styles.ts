import {
  ELementDimensionMap,
  ElementBorderRadius,
  ElementDimension,
  ElementFlexStyleProps,
  ElementMargin,
  ElementPadding,
  ElementPositionMap,
  ElementTextStyleProps,
  ElementViewStyles,
  KeyOfStylePalette,
} from './styleTypes';

export const styles: Record<
  | keyof ELementDimensionMap<ElementPadding | ElementMargin | ElementDimension>
  | ElementBorderRadius
  | keyof ElementPositionMap
  | keyof ElementTextStyleProps
  | keyof ElementFlexStyleProps
  | keyof ElementViewStyles,
  KeyOfStylePalette
> = {
  p: 'padding',
  px: 'paddingHorizontal',
  py: 'paddingVertical',
  ps: 'paddingStart',
  pe: 'paddingEnd',
  pt: 'paddingTop',
  pb: 'paddingBottom',
  m: 'margin',
  mx: 'marginHorizontal',
  my: 'marginVertical',
  ms: 'marginStart',
  me: 'marginEnd',
  mt: 'marginTop',
  mb: 'marginBottom',
  r: 'borderRadius',
  ret: 'borderTopEndRadius',
  rlt: 'borderTopLeftRadius',
  ree: 'borderEndEndRadius',
  rrt: 'borderTopRightRadius',
  rse: 'borderStartEndRadius',
  res: 'borderStartEndRadius',
  rst: 'borderTopStartRadius',
  reb: 'borderBottomEndRadius',
  rlb: 'borderBottomLeftRadius',
  rss: 'borderStartStartRadius',
  rrb: 'borderBottomRightRadius',
  rsb: 'borderBottomStartRadius',
  pos: 'position',
  posB: 'bottom',
  posL: 'left',
  posR: 'right',
  posT: 'top',
  index: 'zIndex',
  w: 'width',
  h: 'height',
  minW: 'minWidth',
  minH: 'minHeight',
  maxW: 'maxWidth',
  maxH: 'maxHeight',
  color: 'color',
  family: 'fontFamily',
  size: 'fontSize',
  style: 'fontStyle',
  weight: 'fontWeight',
  lSpacing: 'letterSpacing',
  lHeight: 'lineHeight',
  dLine: 'textDecorationLine',
  dStyle: 'textDecorationStyle',
  dColor: 'textDecorationColor',
  sColor: 'shadowColor',
  sOpacity: 'shadowOpacity',
  sOffset: 'shadowOffset',
  sRadius: 'shadowRadius',
  transform: 'textTransform',
  select: 'userSelect',
  align: 'alignContent',
  content: 'justifyContent',
  items: 'alignItems',
  self: 'alignSelf',
  ratio: 'aspectRatio',
  d: 'display',
  end: 'end',
  f: 'flex',
  fBasis: 'flexBasis',
  fDirection: 'flexDirection',
  rGap: 'rowGap',
  gap: 'gap',
  cGap: 'columnGap',
  fGrow: 'flexGrow',
  fShrink: 'flexShrink',
  wrap: 'flexWrap',
  bVisibility: 'backfaceVisibility',
  bg: 'backgroundColor',
  o: 'opacity',
  e: 'elevation',
  pEvents: 'pointerEvents',
  c: 'cursor',
  bColor: 'borderColor',
  bWidth: 'borderWidth',
  overflow: 'overflow',
};
