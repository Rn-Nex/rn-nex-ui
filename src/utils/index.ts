export * from './styleGenerator';
export * from './randomId';
export * from './utils';
export * from './dimensions';
