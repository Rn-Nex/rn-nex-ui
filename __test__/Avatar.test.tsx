import { render } from '@testing-library/react-native';
import React from 'react';
import { Image, StyleProp, ViewStyle } from 'react-native';
import { Avatar } from '../src';
import { BaseStyles } from '../src/libraries/style/styleTypes';

describe('Avatar Component', () => {
  const ref = React.createRef<Image>();

  const mockTestId = 'mock_avatar_test_id';
  const mockAvatarUrl = 'avatar.jpg';

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should match the snapshot with  props', () => {
    const { toJSON } = render(<Avatar source={{ uri: mockAvatarUrl }} />);
    expect(toJSON).toMatchSnapshot();
  });

  it('renders correctly with default props', () => {
    const { getByTestId } = render(<Avatar testID={mockTestId} source={{ uri: mockAvatarUrl }} />);

    const avatar = getByTestId(mockTestId);
    expect(avatar).toBeTruthy();

    expect(avatar.props.source).toEqual(expect.objectContaining({ uri: mockAvatarUrl }));
  });

  it('passes size prop correctly to the Image component', () => {
    const { getByTestId } = render(<Avatar testID={mockAvatarUrl} source={{ uri: mockAvatarUrl }} size={50} />);
    const avatar = getByTestId(mockAvatarUrl);
    expect(avatar.props.style).toContainEqual({ width: 50, height: 50 });
  });

  it('applies sx styles correctly', () => {
    const sx: BaseStyles = { bg: 'blue', bWidth: 2 };
    const { getByTestId } = render(<Avatar testID={mockAvatarUrl} source={{ uri: mockAvatarUrl }} sx={sx} />);

    const avatar = getByTestId(mockAvatarUrl);

    const expectedStyles: StyleProp<ViewStyle> = [{ backgroundColor: 'blue', borderWidth: 2 }];
    expect(avatar.props.style).toEqual(expect.arrayContaining(expectedStyles));
  });

  it('forwards ref to the Image component', () => {
    render(<Avatar ref={ref} source={{ uri: mockAvatarUrl }} />);
    expect(ref.current).toBeInstanceOf(Image);
  });

  it('applies the variation prop correctly for different variants', () => {
    const { getByTestId, rerender } = render(<Avatar testID={mockTestId} source={{ uri: mockAvatarUrl }} variation="rounded" />);

    const avatar = getByTestId(mockTestId);
    expect(avatar.props.style).toContainEqual({ borderRadius: 100 });

    rerender(<Avatar testID={mockTestId} source={{ uri: mockAvatarUrl }} variation="square" />);
    expect(avatar.props.style).toContainEqual({ borderRadius: 0 });

    rerender(<Avatar testID={mockTestId} source={{ uri: mockAvatarUrl }} variation="rounded-lg" />);
    expect(avatar.props.style).toContainEqual({ borderRadius: 40 });
  });
});
