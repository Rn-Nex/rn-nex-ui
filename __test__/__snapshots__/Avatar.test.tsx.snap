// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar Component should match the snapshot with  props 1`] = `[Function]`;
